pkg/log/log.go

```diff
@@ -0,0 +1,89 @@
+package log
+
+import (
+	"flag"
+	"fmt"
+	"io"
+
+	"github.com/operator-framework/operator-sdk/pkg/log/zap"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+
+	"github.com/gin-gonic/gin"
+	"github.com/go-logr/logr"
+	"github.com/spf13/pflag"
+)
+
+var (
+	log logr.Logger
+)
+
+// InitializeLogger initializes the logger.
+func InitializeLogger(withName string) {
+	log = logf.Log.WithName(withName)
+
+	zapFlagSet := pflag.NewFlagSet("zap", pflag.ExitOnError)
+
+	// Add the zap logger flag set to the CLI. The flag set must
+	// be added before calling pflag.Parse().
+	pflag.CommandLine.AddFlagSet(zapFlagSet)
+
+	// Add flags registered by imported packages (e.g. glog and
+	// controller-runtime)
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+
+	pflag.Parse()
+
+	// Use a zap logr.Logger implementation. If none of the zap
+	// flags are configured (or if the zap flag set is not being
+	// used), this defaults to a production zap logger.
+	//
+	// The logger instantiated here can be changed to any logger
+	// implementing the logr.Logger interface. This logger will
+	// be propagated through the whole operator, generating
+	// uniform and structured logs.
+	logf.SetLogger(zap.Logger())
+}
+
+func SetOutput(out io.Writer, isTestingMode bool, withName string) {
+	log = logf.ZapLoggerTo(out, isTestingMode).WithName(withName)
+}
+
+// Logger returns the current logger object.
+func Logger() logr.Logger {
+	return log
+}
+
+// Info logs are used for non-error messages. It will log a message with
+// the given key/value pairs as context.
+func Info(ctx *gin.Context, msg string, v ...interface{}) {
```

Comment:
I see your point with regards to API consistency, @alexeykazakov. I like this way of passing args compared to the actual k8s logger (which is a bit counter-intuitive and error prone), but if we don't have a use for these `key`, it's going to feel like an overcomplicated API (compared to logrus's `Infof`, `Warnf`, etc..) for example. Yet, having log messages as "constants" and all contextual info in separate parts of the log output is valuable when we need to search for occurrences of an error for example.

Reply:
So, first of all, if we don't need to implement some specific interface then I would definitely introduce `Infof` so we don't need to use `Sprintf` every time when we need to format the message (as we can see all over the place in this PR). When we introduce `log.Infof(ctx, msg string, args ...string)` then we cannot use the slice of strings for passing the parameters at the end of the method - we need to put it somewhere else.
The thing that I don't like in logrus style is when I don't need to specify any parameter, then I need to pass an empty map in the function anyway. Thus, I would propose having a method `log.WithValues(keysAndValues ...interface{}) Logger` that returns a logger with the key-values set and then we could call either `Error` or `Info` on that Logger instance. 

So, to sum up - let's have functions:
```go
func Infof(ctx, msg string, args ...string) {}
func Errorf(ctx, err error, msg string, args ...string) {}
func WithValues(keysAndValues ...interface{}) log.Logger {}
```
so we can then use it in this way:
```go
log.Infof(ctx, "the action %s has been processed, action)
log.Errorf(ctx, err, "the action %s has failed, action)
```
or in this way:
```go
log.WithValues(map[string]interface{}{
	"name": "john",
}).Errorf(ctx, err, "the %s action has failed", action)
```
or in this way

```go
logger := log.WithValues(map[string]interface{}{
	"name": "john",
})
if err != nil {
    logger.Errorf(ctx, err, "the %s action has failed", action)
    return err
}
logger.Infof(ctx, "the action %s has been processed, action)
```

Reply:
@MatousJobanek sounds good, this looks nice. will make this change

Reply:
thanks @tinakurian 
btw, you could also add function:
```go
func WithValue(key, value string) log.Logger {}
```
in case you wanted to add only one key-value pair

Reply:
> When we introduce log.Infof(ctx, msg string, args ...string) then we cannot use the slice of strings for passing the parameters at the end of the method - we need to put it somewhere else.

Yeah but this is why I mentioned above that we could just pass `nil` when we don't have any keys-values": `log.Info(ctx, nil, "some info message about %s", arg)`. But I like `WithValues(map[string]interface{}) log.Logger` better!
But we still should use `map[string]interface{}` as the param instead of `keysAndValues ...interface{}` IMO. Exactly as you did it in your examples. It will be less error prone.

Reply:
i think the `WithValue` func is not necessary. You can just add a single key/value pair to WithValues if you want to.. however, i still need to change the definition to accept a map..will do that change soon..

Reply:
fixed here: f1dfbdc0e3f6ec6ffb4d94d71281a675eb84d12f